### DISCUSSION TITLE
How to combine several plugins inside next.config.js file?

### DISCUSSION THREAD

**Discussion by sachithrrra** [vercel/next.js#40710] [Help] [answered]
In my next.config.js file, I want to export two module exports. But I can't do that without combining them. Is there a way to combine the two module.exports?

    const withPWA = require("next-pwa");
    const withMDX = require("@next/mdx");
    
    /** @type {import('next').NextConfig} */
    const nextConfig = {
      reactStrictMode: true,
      i18n: {
        locales: ["en"],
        defaultLocale: "en",
      },
      swcMinify: true,
      compiler: {
        removeConsole: process.env.NODE_ENV !== "development",
      },
      //pwa config
      pwa: {
        dest: "public",
        disable: process.env.NODE_ENV === "development",
        register: true,
      },
      //mdx config
      extension: /\.mdx?$/,
      options: {
        remarkPlugins: [],
        rehypePlugins: [],
        // If you use `MDXProvider`, uncomment the following line.
        // providerImportSource: "@mdx-js/react",
      },
      pageExtensions: ["ts", "tsx", "js", "jsx", "md", "mdx"],
    };
    
    module.exports = () => {
      const plugins = [withPWA, withMDX];
      const config = plugins.reduce((acc, next) => next(acc), {
        ...nextConfig,
      });
      return config;
    };

**Comment by icyJoseph** [comment, marked as answer]
Node.js CommonJS modules can only have one module.exports call, or rather, the last one is the one that applies.

Could you try with:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  i18n: {
    locales: ["en"],
    defaultLocale: "en",
  },
  swcMinify: true,
  compiler: {
    removeConsole: process.env.NODE_ENV !== "development",
  },
  pageExtensions: ["ts", "tsx", "js", "jsx", "md", "mdx"],
};

const pwaConfig = {
  dest: "public",
  disable: process.env.NODE_ENV === "development",
  register: true,
};

const mdxConfig = {
  extension: /.mdx?$/,
  options: {
    remarkPlugins: [],
    rehypePlugins: [],
    // If you use MDXProvider, uncomment the following line.
    // providerImportSource: "@mdx-js/react",
  },
};

const withPWA = require("next-pwa")(pwaConfig);
const withMDX = require("@next/mdx")(mdxConfig);

module.exports = withMDX(withPWA(nextConfig));
```

## Edit

OP is using `5.5.4` which does not comply with the recommended configuration pattern. The snippet above works for `5.6.0` and later versions. To make it work with earlier versions do:

```js
const withPWA = require("next-pwa");
const withMDX = require("@next/mdx")(mdxConfig);

module.exports = withMDX(withPWA({ ...nextConfig, pwa: pwaConfig }));
```

However, that'll log a bunch of warnings.

**Reply by sachithrrra** [reply]
hey @icyJoseph I get this error message with this config

    PS D:\Personal\Projects\MyProject\MyProject> npm run dev
    
    > dev
    > next dev
    
    ready - started server on 0.0.0.0:3000, url: http://localhost:3000
    info  - Loaded env from D:\Personal\Projects\MyProject\MyProject\.env.local
    error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
    TypeError: withPWA is not a function
        at Object.<anonymous> (D:\Personal\Projects\MyProject\MyProject\next.config.js:96:26)
        at Module._compile (internal/modules/cjs/loader.js:1063:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
        at Module.load (internal/modules/cjs/loader.js:928:32)
        at Function.Module._load (internal/modules/cjs/loader.js:769:14)
        at ModuleWrap.<anonymous> (internal/modules/esm/translators.js:199:29)
        at ModuleJob.run (internal/modules/esm/module_job.js:152:23)
        at async Loader.import (internal/modules/esm/loader.js:166:24)
        at async Object.loadConfig [as default] (D:\Personal\Projects\MyProject\MyProject\node_modules\next\dist\server\config.js:68:36)

**Reply by icyJoseph** [reply]
Mmm have you installed `next-pwa`?  Sad to say that it works on my machine. Got a repository to share? Or could you share what are you exactly doing?

**Reply by sachithrrra** [reply]
Yes. it's installed. I want to use both of these plugins in the NextJS app. `next-pwa` and `next/mdx`

**Reply by icyJoseph** [reply]
Like I said, the snippet I posted worked when I tried it. Would you mind sharing a link to a repository? Or at least the contents of your next config file? 

What next version have you got and other factors might be at play here.

**Reply by sachithrrra** [reply]
This is the complete `next.config.js` file

    /** @type {import('next').NextConfig} */
    const nextConfig = {
      reactStrictMode: true,
      i18n: {
        locales: ["en"],
        defaultLocale: "en",
      },
      swcMinify: true,
      compiler: {
        removeConsole: process.env.NODE_ENV !== "development",
      },
      pageExtensions: ["ts", "tsx", "js", "jsx", "md", "mdx"],
    };
    
    const pwaConfig = {
      dest: "public",
      disable: process.env.NODE_ENV === "development",
      register: true,
    };
    
    const mdxConfig = {
      extension: /.mdx?$/,
      options: {
        remarkPlugins: [],
        rehypePlugins: [],
        // If you use MDXProvider, uncomment the following line.
        // providerImportSource: "@mdx-js/react",
      },
    };
    
    const withPWA = require("next-pwa")(pwaConfig);
    const withMDX = require("@next/mdx")(mdxConfig);
    
    module.exports = withMDX(withPWA(nextConfig));

this is the `package.json` file

    {
    
    "name": "myproject",
    
    "private": true,
    
    "scripts": {
    
    "dev": "next dev",
    
    "build": "next build",
    
    "postbuild": "next-sitemap",
    
    "start": "next start",
    
    "lint": "next lint"
    
    },
    
    "dependencies": {
    
    "@headlessui/react": "^1.6.6",
    
    "axios": "^0.26.1",
    
    "downloadjs": "^1.4.7",
    
    "html-to-image": "^1.9.0",
    
    "lodash": "^4.17.21",
    
    "moment": "^2.29.4",
    
    "next": "^12.3.0",
    
    "next-pwa": "5.5.4",
    
    "next-seo": "^5.4.0",
    
    "next-sitemap": "^3.1.11",
    
    "nextjs-google-analytics": "^1.2.0",
    
    "numeral": "^2.0.6",
    
    "react": "^18.2.0",
    
    "react-dom": "^18.2.0",
    
    "react-select": "^5.3.2"
    
    },
    
    "devDependencies": {
    
    "@tailwindcss/typography": "^0.5.4",
    
    "autoprefixer": "^10.4.4",
    
    "eslint": "8.12.0",
    
    "eslint-config-next": "12.1.4",
    
    "postcss": "^8.4.5",
    
    "postcss-nested": "^5.0.6",
    
    "prettier": "^2.7.1",
    
    "prettier-plugin-tailwindcss": "^0.1.13",
    
    "tailwindcss": "^3.0.23"
    
    }
    
    }

**Reply by icyJoseph** [reply]
Well both @next/mdx and its family of packages are missing, according to https://www.npmjs.com/package/@next/mdx:

```
yarn add @next/mdx @mdx-js/loader @mdx-js/react
```

Or install with NPM if you use that.

The error you got happens because of the next-pwa version you have. I'd recommend taking `5.6.0`, or `npm install next-pwa@latest`, because of some alignment they've done with the `next.config.js` way of working.

Let's be clear here, my snippet above works with `5.6.0` and above. The reason is detailed here by the maintainers of `next-pwa`, https://github.com/shadowwalker/next-pwa/releases/tag/5.6.0.

If you want to stick to versions under, such as `5.5.4` use this:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  i18n: {
    locales: ["en"],
    defaultLocale: "en",
  },
  swcMinify: true,
  compiler: {
    removeConsole: process.env.NODE_ENV !== "development",
  },
  pageExtensions: ["ts", "tsx", "js", "jsx", "md", "mdx"],
};

const pwaConfig = {
  dest: "public",
  disable: process.env.NODE_ENV === "development",
  register: true,
};

const mdxConfig = {
  extension: /.mdx?$/,
  options: {
    remarkPlugins: [],
    rehypePlugins: [],
    // If you use MDXProvider, uncomment the following line.
    // providerImportSource: "@mdx-js/react",
  },
};

const withPWA = require("next-pwa");
const withMDX = require("@next/mdx")(mdxConfig);

module.exports = withMDX(withPWA({ ...nextConfig, pwa: pwaConfig }));
```

However, the above will yield a bunch of warnings to the console.

**Reply by sachithrrra** [reply]
Updated `next-pwa` to `5.6.0` and It's working as expected. Thank you for your time. Really appreciate it.
